File: src/stdlib/flock.ts
/**
 * `flock`: collections.
 *
 * A flock is a group, so `flock` holds the operations that work across whole
 * arrays and maps. Anything that reads naturally as a method (`items.map(f)`)
 * already is one; this module covers the shapes that don't: zipping, grouping,
 * chunking, building maps from pairs.
 */

import { BaaError } from "../diagnostics/diagnostic.ts";
import {
  BaaArray,
  BaaMap,
  BaaRange,
  inspect,
  isMapKey,
  isTruthy,
  typeOf,
} from "../runtime/values.ts";
import type { MapKey, NativeContext, Value } from "../runtime/values.ts";
import { argAny, argArray, argInt, argMap, defineModule, fn } from "./define.ts";

export function createFlock() {
  return defineModule("flock", {
    of: fn(0, Number.MAX_SAFE_INTEGER, "Build an array from the arguments.", (args) =>
      new BaaArray([...args]),
    ),

    repeat: fn(2, 2, "An array with the same value repeated n times.", (args, ctx) => {
      const value = argAny(args, 0);
      const count = argInt("flock.repeat", args, 1, ctx.span);
      if (count < 0) {
        throw BaaError.of("BAA311", ["flock.repeat", "a count of 0 or more", "2", "a negative number"], {
          span: ctx.span,
        });
      }
      return new BaaArray(new Array<Value>(count).fill(value));
    }),

    zip: fn(2, 2, "Pair up two arrays, stopping at the shorter one.", (args, ctx) => {
      const left = argArray("flock.zip", args, 0, ctx.span);
      const right = argArray("flock.zip", args, 1, ctx.span);
      const length = Math.min(left.items.length, right.items.length);
      const out: Value[] = [];
      for (let i = 0; i < length; i++) {
        out.push(new BaaArray([left.items[i]!, right.items[i]!]));
      }
      return new BaaArray(out);
    }),

    chunk: fn(2, 2, "Split an array into chunks of a fixed size.", (args, ctx) => {
      const items = argArray("flock.chunk", args, 0, ctx.span);
      const size = argInt("flock.chunk", args, 1, ctx.span);
      if (size < 1) {
        throw BaaError.of("BAA311", ["flock.chunk", "a size of 1 or more", "2", "a smaller number"], {
          span: ctx.span,
        });
      }
      const out: Value[] = [];
      for (let i = 0; i < items.items.length; i += size) {
        out.push(new BaaArray(items.items.slice(i, i + size)));
      }
      return new BaaArray(out);
    }),

    group_by: fn(2, 2, "Group items into a map keyed by fn(item).", (args, ctx) => {
      const items = argArray("flock.group_by", args, 0, ctx.span);
      const keyFn = argAny(args, 1);
      const groups = new Map<MapKey, Value>();
      for (const item of items.items) {
        const key = ctx.call(keyFn, [item], ctx.span);
        if (!isMapKey(key)) {
          throw BaaError.of(
            "BAA311",
            ["flock.group_by", "a string, number, bool or nil key", "2", typeOf(key)],
            { span: ctx.span, note: "the key function returned something unusable" },
          );
        }
        const bucket = groups.get(key);
        if (bucket instanceof BaaArray) bucket.items.push(item);
        else groups.set(key, new BaaArray([item]));
      }
      return new BaaMap(groups);
    }),

    partition: fn(2, 2, "Split into [matching, rest] using a predicate.", (args, ctx) => {
      const items = argArray("flock.partition", args, 0, ctx.span);
      const predicate = argAny(args, 1);
      const yes: Value[] = [];
      const no: Value[] = [];
      for (const item of items.items) {
        (isTruthy(ctx.call(predicate, [item], ctx.span)) ? yes : no).push(item);
      }
      return new BaaArray([new BaaArray(yes), new BaaArray(no)]);
    }),

    sort_by: fn(2, 2, "Sorted copy, ordered by the key fn(item) returns.", (args, ctx) => {
      const items = argArray("flock.sort_by", args, 0, ctx.span);
      const keyFn = argAny(args, 1);
      const decorated = items.items.map((item) => ({
        item,
        key: ctx.call(keyFn, [item], ctx.span),
      }));
      decorated.sort((a, b) => {
        if (typeof a.key === "number" && typeof b.key === "number") return a.key - b.key;
        if (typeof a.key === "string" && typeof b.key === "string") {
          return a.key < b.key ? -1 : a.key > b.key ? 1 : 0;
        }
        throw BaaError.of("BAA302", ["compare", typeOf(a.key), typeOf(b.key)], {
          span: ctx.span,
          note: "sort keys must be all numbers or all strings",
        });
      });
      return new BaaArray(decorated.map((entry) => entry.item));
    }),

    min_by: fn(2, 2, "Item with the smallest fn(item), or nil when empty.", (args, ctx) =>
      pickBy(args, ctx, -1),
    ),

    max_by: fn(2, 2, "Item with the largest fn(item), or nil when empty.", (args, ctx) =>
      pickBy(args, ctx, 1),
    ),

    to_map: fn(1, 1, "Build a map from an array of [key, value] pairs.", (args, ctx) => {
      const pairs = argArray("flock.to_map", args, 0, ctx.span);
      const entries = new Map<MapKey, Value>();
      for (const pair of pairs.items) {
        if (!(pair instanceof BaaArray) || pair.items.length !== 2) {
          throw BaaError.of("BAA311", ["flock.to_map", "an array of [key, value] pairs", "1", inspect(pair)], {
            span: ctx.span,
          });
        }
        const key = pair.items[0]!;
        if (!isMapKey(key)) {
          throw BaaError.of("BAA311", ["flock.to_map", "a usable map key", "1", typeOf(key)], {
            span: ctx.span,
          });
        }
        entries.set(key, pair.items[1]!);
      }
      return new BaaMap(entries);
    }),

    from_keys: fn(2, 2, "Build a map giving every key the same value.", (args, ctx) => {
      const keys = argArray("flock.from_keys", args, 0, ctx.span);
      const value = argAny(args, 1);
      const entries = new Map<MapKey, Value>();
      for (const key of keys.items) {
        if (!isMapKey(key)) {
          throw BaaError.of("BAA311", ["flock.from_keys", "usable map keys", "1", typeOf(key)], {
            span: ctx.span,
          });
        }
        entries.set(key, value);
      }
      return new BaaMap(entries);
    }),

    invert: fn(1, 1, "Swap a map's keys and values.", (args, ctx) => {
      const map = argMap("flock.invert", args, 0, ctx.span);
      const entries = new Map<MapKey, Value>();
      for (const [key, value] of map.entries) {
        if (!isMapKey(value)) {
          throw BaaError.of("BAA311", ["flock.invert", "values usable as keys", "1", typeOf(value)], {
            span: ctx.span,
          });
        }
        entries.set(value, key);
      }
      return new BaaMap(entries);
    }),

    range: fn(1, 3, "An array of numbers: range(end), range(start, end[, step]).", (args, ctx) => {
      const first = argInt("flock.range", args, 0, ctx.span);
      const start = args.length === 1 ? 0 : first;
      const end = args.length === 1 ? first : argInt("flock.range", args, 1, ctx.span);
      const step = args.length > 2 ? argInt("flock.range", args, 2, ctx.span) : 1;
      if (step === 0) {
        throw BaaError.of("BAA311", ["flock.range", "a non-zero step", "3", "zero"], {
          span: ctx.span,
        });
      }
      const out: Value[] = [];
      if (step > 0) for (let i = start; i < end; i += step) out.push(i);
      else for (let i = start; i > end; i += step) out.push(i);
      return new BaaArray(out);
    }),

    to_array: fn(1, 1, "Turn a range, string or map into an array.", (args, ctx) => {
      const value = argAny(args, 0);
      if (value instanceof BaaArray) return new BaaArray([...value.items]);
      if (value instanceof BaaRange) return new BaaArray([...value.values()]);
      if (typeof value === "string") return new BaaArray([...value]);
      if (value instanceof BaaMap) {
        return new BaaArray(
          [...value.entries].map(([key, entry]) => new BaaArray([key, entry])),
        );
      }
      throw BaaError.of("BAA311", ["flock.to_array", "an array, range, string or map", "1", typeOf(value)], {
        span: ctx.span,
      });
    }),
  });
}

function pickBy(args: Value[], ctx: NativeContext, direction: 1 | -1): Value {
  const items = argArray(direction === 1 ? "flock.max_by" : "flock.min_by", args, 0, ctx.span);
  const keyFn = argAny(args, 1);
  let best: Value = null;
  let bestKey: Value = null;
  for (const item of items.items) {
    const key = ctx.call(keyFn, [item], ctx.span);
    if (typeof key !== "number" && typeof key !== "string") {
      throw BaaError.of("BAA311", ["flock.min_by/max_by", "a number or string key", "2", typeOf(key)], {
        span: ctx.span,
      });
    }
    if (bestKey === null || compare(key, bestKey) * direction > 0) {
      best = item;
      bestKey = key;
    }
  }
  return best;
}

function compare(a: Value, b: Value): number {
  if (typeof a === "number" && typeof b === "number") return a - b;
  return String(a) < String(b) ? -1 : String(a) > String(b) ? 1 : 0;
}

File: rust/crates/baa-native/src/stdlib/flock.rs
//! `flock`: collections. A port of `src/stdlib/flock.ts`.
//!
//! Every function takes a collection and returns a new one; nothing here
//! mutates its argument. That is the same contract the reference has, and it
//! is what makes `flock.sort_by(sheep, age)` safe to call on something another
//! part of the program is still holding.

use std::rc::Rc;

use crate::ast::Span;
use crate::interp::{Flow, Interpreter, Res, MAX_SIZE};
use crate::number;
use crate::value::{BaaMap, MapKey, Module, Native, Value};

use super::{module_of, need_array, need_int};

const FUNCTIONS: &[(&str, usize, usize)] = &[
    ("of", 0, usize::MAX),
    ("repeat", 2, 2),
    ("zip", 2, 2),
    ("chunk", 2, 2),
    ("group_by", 2, 2),
    ("partition", 2, 2),
    ("sort_by", 2, 2),
    ("min_by", 2, 2),
    ("max_by", 2, 2),
    ("to_map", 1, 1),
    ("from_keys", 2, 2),
    ("invert", 1, 1),
    ("range", 1, 3),
    ("to_array", 1, 1),
];

pub fn module() -> Rc<Module> {
    let exports = FUNCTIONS
        .iter()
        .map(|(name, min, max)| {
            (
                *name,
                Value::Native(Rc::new(Native {
                    name: format!("flock.{name}"),
                    method: name,
                    min_args: *min,
                    max_args: *max,
                    call,
                    receiver: None,
                })),
            )
        })
        .collect();
    module_of("flock", exports)
}

fn check_size(interp: &Interpreter, name: &str, count: f64, span: Span) -> Res<()> {
    if !count.is_finite() || count > MAX_SIZE as f64 {
        return Err(Flow::Err(
            interp
                .error(
                    "BAA312",
                    vec![name.to_string(), number::format(count), number::format(MAX_SIZE as f64)],
                    span,
                )
                .with_note("too many at once"),
        ));
    }
    Ok(())
}

/// Compares two sort keys, which must be all numbers or all strings.
fn compare_keys(interp: &Interpreter, a: &Value, b: &Value, span: Span) -> Res<std::cmp::Ordering> {
    match (a, b) {
        (Value::Number(x), Value::Number(y)) => Ok(x.partial_cmp(y).unwrap_or(std::cmp::Ordering::Equal)),
        (Value::Str(x), Value::Str(y)) => Ok(x.cmp(y)),
        _ => Err(Flow::Err(
            interp
                .error("BAA302", vec!["compare".into(), a.describe(), b.describe()], span)
                .with_note("sort keys must be all numbers or all strings"),
        )),
    }
}

fn call(interp: &mut Interpreter, native: &Native, args: Vec<Value>, span: Span) -> Res<Value> {
    let name = native.name.clone();
    Ok(match native.method {
        "of" => Value::array(args),
        "repeat" => {
            let count = need_int(interp, &name, &args, 1, span)?;
            check_size(interp, &name, count, span)?;
            let value = args[0].clone();
            Value::array(vec![value; count.max(0.0) as usize])
        }
        "zip" => {
            let left = need_array(interp, &name, &args, 0, span)?;
            let right = need_array(interp, &name, &args, 1, span)?;
            Value::array(
                left.into_iter()
                    .zip(right)
                    .map(|(a, b)| Value::array(vec![a, b]))
                    .collect(),
            )
        }
        "chunk" => {
            let items = need_array(interp, &name, &args, 0, span)?;
            let size = need_int(interp, &name, &args, 1, span)?;
            if size < 1.0 {
                return interp.fail(
                    "BAA311",
                    vec![name, "a size of 1 or more".into(), "2".into(), number::format(size)],
                    span,
                );
            }
            Value::array(
                items
                    .chunks(size as usize)
                    .map(|chunk| Value::array(chunk.to_vec()))
                    .collect(),
            )
        }
        "group_by" => {
            let items = need_array(interp, &name, &args, 0, span)?;
            let key_fn = args[1].clone();
            let mut groups = BaaMap::new();
            for item in items {
                let key_value = interp.call_callback(&key_fn, vec![item.clone()], span)?;
                let Some(key) = MapKey::from_value(&key_value) else {
                    return interp.fail(
                        "BAA311",
                        vec![name, "a key the map accepts".into(), "2".into(), key_value.describe()],
                        span,
                    );
                };
                match groups.get(&key) {
                    Some(Value::Array(existing)) => existing.borrow_mut().push(item),
                    _ => groups.set(key, Value::array(vec![item])),
                }
            }
            Value::map(groups)
        }
        "partition" => {
            let items = need_array(interp, &name, &args, 0, span)?;
            let test = args[1].clone();
            let (mut yes, mut no) = (Vec::new(), Vec::new());
            for item in items {
                if interp.call_callback(&test, vec![item.clone()], span)?.truthy() {
                    yes.push(item);
                } else {
                    no.push(item);
                }
            }
            Value::array(vec![Value::array(yes), Value::array(no)])
        }
        "sort_by" => {
            let items = need_array(interp, &name, &args, 0, span)?;
            let key_fn = args[1].clone();
            // Decorate, sort, undecorate: the key function runs once per item
            // rather than twice per comparison, which matters when it is Baa
            // code rather than a field read.
            let mut decorated = Vec::with_capacity(items.len());
            for item in items {
                let key = interp.call_callback(&key_fn, vec![item.clone()], span)?;
                decorated.push((key, item));
            }
            let mut failure = None;
            decorated.sort_by(|a, b| match compare_keys(interp, &a.0, &b.0, span) {
                Ok(ordering) => ordering,
                Err(flow) => {
                    if failure.is_none() {
                        failure = Some(flow);
                    }
                    std::cmp::Ordering::Equal
                }
            });
            if let Some(flow) = failure {
                return Err(flow);
            }
            Value::array(decorated.into_iter().map(|(_, item)| item).collect())
        }
        "min_by" | "max_by" => {
            let items = need_array(interp, &name, &args, 0, span)?;
            let key_fn = args[1].clone();
            let want_greater = native.method == "max_by";
            let mut best: Option<(Value, Value)> = None;
            for item in items {
                let key = interp.call_callback(&key_fn, vec![item.clone()], span)?;
                match &best {
                    None => best = Some((key, item)),
                    Some((best_key, _)) => {
                        let ordering = compare_keys(interp, &key, best_key, span)?;
                        if (ordering == std::cmp::Ordering::Greater) == want_greater
                            && ordering != std::cmp::Ordering::Equal
                        {
                            best = Some((key, item));
                        }
                    }
                }
            }
            best.map(|(_, item)| item).unwrap_or(Value::Nil)
        }
        "to_map" => {
            let items = need_array(interp, &name, &args, 0, span)?;
            let mut map = BaaMap::new();
            for (index, pair) in items.iter().enumerate() {
                let Value::Array(entry) = pair else {
                    return interp.fail(
                        "BAA311",
                        vec![name, "an array of [key, value] pairs".into(), "1".into(), pair.describe()],
                        span,
                    );
                };
                let entry = entry.borrow();
                let key_value = entry.first().cloned().unwrap_or(Value::Nil);
                let Some(key) = MapKey::from_value(&key_value) else {
                    return interp.fail(
                        "BAA311",
                        vec![
                            name,
                            format!("a key the map accepts at position {index}"),
                            "1".into(),
                            key_value.describe(),
                        ],
                        span,
                    );
                };
                map.set(key, entry.get(1).cloned().unwrap_or(Value::Nil));
            }
            Value::map(map)
        }
        "from_keys" => {
            let keys = need_array(interp, &name, &args, 0, span)?;
            let value_fn = args[1].clone();
            let mut map = BaaMap::new();
            for key_value in keys {
                let Some(key) = MapKey::from_value(&key_value) else {
                    return interp.fail(
                        "BAA311",
                        vec![name, "keys the map accepts".into(), "1".into(), key_value.describe()],
                        span,
                    );
                };
                let value = interp.call_callback(&value_fn, vec![key_value], span)?;
                map.set(key, value);
            }
            Value::map(map)
        }
        "invert" => {
            let Some(Value::Map(source)) = args.first() else {
                return interp.fail(
                    "BAA311",
                    vec![name, "a map".into(), "1".into(), args.first().cloned().unwrap_or(Value::Nil).describe()],
                    span,
                );
            };
            let entries: Vec<(MapKey, Value)> = source.borrow().iter().cloned().collect();
            let mut inverted = BaaMap::new();
            for (key, value) in entries {
                let Some(new_key) = MapKey::from_value(&value) else {
                    return interp.fail(
                        "BAA311",
                        vec![name, "values the map accepts as keys".into(), "1".into(), value.describe()],
                        span,
                    );
                };
                inverted.set(new_key, key.to_value());
            }
            Value::map(inverted)
        }
        "range" => {
            let first = need_int(interp, &name, &args, 0, span)?;
            let (start, end) = if args.len() == 1 {
                (0.0, first)
            } else {
                (first, need_int(interp, &name, &args, 1, span)?)
            };
            let step = if args.len() > 2 { need_int(interp, &name, &args, 2, span)? } else { 1.0 };
            if step == 0.0 {
                return interp.fail(
                    "BAA311",
                    vec![name, "a non-zero step".into(), "3".into(), "zero".into()],
                    span,
                );
            }
            check_size(interp, &name, ((end - start) / step).ceil().max(0.0), span)?;
            let mut out = Vec::new();
            let mut at = start;
            if step > 0.0 {
                while at < end {
                    out.push(Value::Number(at));
                    at += step;
                }
            } else {
                while at > end {
                    out.push(Value::Number(at));
                    at += step;
                }
            }
            Value::array(out)
        }
        "to_array" => match args.first() {
            Some(Value::Array(items)) => Value::array(items.borrow().clone()),
            Some(Value::Range(range)) => {
                check_size(interp, &name, range.length(), span)?;
                Value::array(range.values().into_iter().map(Value::Number).collect())
            }
            Some(Value::Str(text)) => {
                Value::array(text.chars().map(|c| Value::str(c.to_string())).collect())
            }
            Some(Value::Map(map)) => Value::array(
                map.borrow()
                    .iter()
                    .map(|(key, value)| Value::array(vec![key.to_value(), value.clone()]))
                    .collect(),
            ),
            other => {
                return interp.fail(
                    "BAA311",
                    vec![
                        name,
                        "an array, range, string or map".into(),
                        "1".into(),
                        other.cloned().unwrap_or(Value::Nil).describe(),
                    ],
                    span,
                )
            }
        },
        _ => Value::Nil,
    })
}
